src/app-layer-expectation.c

```diff
@@ -0,0 +1,287 @@
+    IPPair *ipp;
+    Address ip_src, ip_dst;
+
+    if (exp == NULL)
+        return -1;
+
+    exp->sp = src;
+    exp->dp = dst;
+    exp->alproto = alproto;
+    exp->ts = f->lastts;
+    exp->data = data;
+    exp->direction = direction;
+
+    if (GetFlowAddresses(f, &ip_src, &ip_dst) == -1)
+        goto error;
+    if (direction & STREAM_TOSERVER) {
+        ipp = IPPairGetIPPairFromHash(&ip_src, &ip_dst);
+    } else {
+        ipp = IPPairGetIPPairFromHash(&ip_dst, &ip_src);
+    }
+    if (ipp == NULL)
+        goto error;
+
+    iexp = IPPairGetStorageById(ipp, expectation_id);
+    exp->next = iexp;
+    IPPairSetStorageById(ipp, expectation_id, exp);
+
+    SC_ATOMIC_ADD(expectation_count, 1);
+    IPPairUnlock(ipp);
+    return 0;
+
+error:
+    if (exp != NULL)
+        SCFree(exp);
+    return -1;
+}
+
+static Expectation *AppLayerExpectationGet(Flow *f, int direction, IPPair **ipp)
+{
+    Address ip_src, ip_dst;
+    if (GetFlowAddresses(f, &ip_src, &ip_dst) == -1)
+        return NULL;
+    if (direction & STREAM_TOSERVER) {
+        *ipp = IPPairLookupIPPairFromHash(&ip_src, &ip_dst);
+    } else {
+        *ipp = IPPairLookupIPPairFromHash(&ip_dst, &ip_src);
+    }
+    if (*ipp == NULL) {
+        return NULL;
+    }
+
+    return IPPairGetStorageById(*ipp, expectation_id);
+}
+
+/**
+ * Return Flow storage identifier corresponding to expectation data
+ *
+ * \return expectation data identifier
+ */
+int AppLayerExpectationGetDataId(void)
+{
+    return expectation_data_id;
+}
+
+static Expectation * RemoveExpectationAndGetNext(IPPair *ipp,
+                                Expectation *pexp, Expectation *exp,
+                                Expectation *lexp)
+{
+    SC_ATOMIC_SUB(expectation_count, 1);
+    if (pexp == NULL) {
+        IPPairSetStorageById(ipp, expectation_id, lexp);
+    } else {
+        pexp->next = lexp;
+    }
+    if (exp->data)
+        SCFree(exp->data);
+    SCFree(exp);
+    return lexp;
+}
+
+/**
+ * Function doing a lookup in expectation list
+ *
+ * \return an AppProto value if found
+ * \return ALPROTO_UNKNOWN if not found
+ */
+AppProto AppLayerExpectationLookup(Flow *f, int direction)
+{
+    AppProto alproto = ALPROTO_UNKNOWN;
+    IPPair *ipp = NULL;
+    Expectation *lexp = NULL;
+    Expectation *pexp = NULL;
+
+    int x = SC_ATOMIC_GET(expectation_count);
+    if (x == 0) {
+        return ALPROTO_UNKNOWN;
+    }
+
+    Expectation *exp = AppLayerExpectationGet(f, direction, &ipp);
+    time_t ctime = time(NULL);
+
+    if (exp == NULL)
+        goto out;
+
+    pexp = NULL;
+    while (exp) {
+        lexp = exp->next;
+        if ( (exp->direction & direction) &&
+             ((exp->sp == 0) || (exp->sp == f->sp)) &&
+             ((exp->dp == 0) || (exp->dp == f->dp))) {
+            alproto = exp->alproto;
+            if (FlowSetStorageById(f, expectation_data_id, exp->data) != 0) {
+                SCLogError(SC_ERR_UNKNOWN_VALUE, "Unable to set flow storage");
```

Comment:
error print in packet path

Reply:
also, is this a memleak? if data is not stored in flow (error) exp->data is still just set to NULL

Reply:
Yes potentially but the set storage funciton should only fails in case the storage is not correctly setup so this is a programmation error if it occurs.
Should we keep the SCLogError in this case ?

A fix would be to probide during the creation of the expectation a cleaning function that could be used to free the structure. I found this a bit overkill for something that should only be reached in case of implementation error.